Run a range of migrations

winklertomas opened the pull request:
### Motivation

Which issue does this fix? Fixes [`issue`](https://github.com/Kentico/kontent-migrations-boilerplate/issues/11)

### Checklist

- [x] Code follows coding conventions held in this repo
- [x] Automated tests have been added
- [x] Tests are passing
- [x] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Try to run a range of migrations

Simply007 commented on src/utils/migrationUtils.ts:
```diff
@@ -152,15 +152,17 @@ export const loadModule = async (migrationFile: string): Promise<MigrationModule
 
     for (const file of files) {
         const migrationModule = await loadModule(file.name);
 
```
I might add conditions to check whether the range is within `migrationrange`.

winklertomas replied:
I didn't add the check as the current implementation is able to handle it too. If there is no migration that matches the specified range and at the same time is not in status.json, the run will end with the message: 'No migrations to run.'

Simply007 commented:
It might be worth describing the argument information in README 
*  mention what happened with `status.json` 
* cover status.json update by test

winklertomas commented:
I've added unit tests to validate the order, however the status.json update itself, in my opinion, doesn't require more coverage. I have similar feelings about the README edits, I've updated it a bit and added an example into the help section, but the new concept of running migrations from a specified range nicely complements the current implementation. It modifies the status.json in the same way, it just writes each migration to status.json to avoid running it again.

I've also added a validity check for the migration order, the current implementation already checks for duplicates, so I think it will be possible to close https://github.com/Kentico/kontent-cli/issues/33 as well. I don't want to check for example that there is no gap in the migration order, it would make deleting migrations unnecessarily difficult.

I will add an example of running a range of migrations to the boilerplate as well

Simply007 commented on src/cmds/migration/run.ts:
```diff
@@ -165,6 +182,17 @@ const runMigrationCommand: yargs.CommandModule = {
+    const from = Number(match[1]);
+    const to = Number(match[2]);
+
+    return from <= to ? [from, to] : null;
```
I might return 
```
{`
  from,
  to
}
```

object

Simply007 commented on tsconfig.json:
```diff
@@ -3,7 +3,8 @@
     "lib": [
-      "esnext"
+      "esnext",
+      "DOM"
```
Why this is neccesary? It seems incorrect since this is a CLI tool.

winklertomas replied:
It is because of this issue https://github.com/axios/axios/issues/4124. It was fixed and merged to the master 8 days ago https://github.com/axios/axios/pull/4229, but it is still not released https://github.com/axios/axios/releases. 

After the release we can upgrade the dependency in https://github.com/Kentico/kontent-core-js/blob/master/package.json

Simply007 commented on src/cmds/migration/run.ts:
```diff
@@ -165,6 +182,17 @@ const runMigrationCommand: yargs.CommandModule = {
+    const from = Number(match[1]);
+    const to = Number(match[2]);
+
+    return from <= to ? [from, to] : null;
```
I might return 
```
{
  from,
  to
}
```

object

Just for readability.

winklertomas replied:
I've introduced new IRange interface, thanks for the tip

Simply007 commented on README.md:
````diff
@@ -112,7 +112,7 @@ The supported commands are divided into groups according to their target, at thi
     ```
 
-* `migration run` - Runs a migration script specified by file name (option `--name <file name>`), or runs multiple migration scripts in the order specified in the migration files (option `--all`).
+* `migration run` - Runs a migration script specified by file name (option `--name <file name>`), or runs multiple migration scripts in the order specified in the migration files (options `--all` or `--range`).
````
We might want to mention that the order does not need to be incremental by 1 (can have gaps)

winklertomas replied:
done, I've also mentioned that order has to be a unique, positive integer or zero

Simply007 commented on src/tests/invalidOrderMigration.test.ts:
```diff
@@ -0,0 +1,66 @@
+        expect(result[2].module.order).toBe('aaa');
+    });
+
+    it('No invalid order is found', () => {
```
We can add test for 2, 4, and 100 order (to showcase the gaps between migrations)

winklertomas replied:
done

Simply007 commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@kentico/kontent-cli",
-  "version": "0.3.0",
+  "version": "0.5.0",
```
Why raise by 2? I guess because 2 features were added, but I think we can still have just 0.4.0 with 2 features.

winklertomas replied:
it's just because I accidentally released 0.4.0-next-1 instead of 0.3.0-next-1 during testing

Simply007 reviewed (commented):
Overall seems OK.
Just a couple of notes -> none of them is super important, but the DOM in tsconfig.json

Simply007 reviewed (approved):
LGTM - the if/else forest in `run.ts` might be refactored, but this is not a part of the task and it is still fairly readable.
